### ISSUE TITLE
New features

### ISSUE THREAD

**Issue by AndBondStyle** [d-ryzhykau/pipenv-lock-pre-commit#3]
I was searching for a good pipenv pre-commit hook, and it looks like yours is the most popular. However, there's some fixes and features I want to be introduced:

1. Switch to a new `pipenv requirements` command instead of `pipenv lock -r`. The latter one will be deprecated:
```
$ pipenv lock -r
Warning: The lock flag -r/--requirements will be deprecated in a future version
of pipenv in favor of the new requirements command. For more info see
https://pipenv.pypa.io/en/latest/advanced/#generating-a-requirements-txt
```
2. Add an argument to disable generation of `requirements-dev.txt` (or move it to separate hook), as it's sometimes not needed
3. Add an extra hook to check if `Pipenv.lock` is up-to-date (`pipenv verify`)
4. Add extra parameters to existing hooks, e.g. `--hash` argument to enable/disable hash generation in requirements.txt
5. Support non-standard `Pipenv` file locations (via hook argument) for repos with multiple sub-projects

If you're okay with that, I'm willing to contribute.

**Comment by d-ryzhykau** [comment]
Hello and sorry for the late answer. For some reason Github decided not to notify me about this issue, I found it once I went by the repository by chance.

I went through the features you proposed and implemented some of them.

> Switch to a new pipenv requirements command instead of pipenv lock -r...

Implemented in version [0.2.0](https://github.com/d-ryzhikov/pipenv-lock-pre-commit/releases/tag/0.2.0).

> Add an argument to disable generation of requirements-dev.txt (or move it to separate hook)...

> Add extra parameters to existing hooks, e.g. --hash argument to enable/disable hash generation in requirements.txt

Implemented in version [0.3.0](https://github.com/d-ryzhikov/pipenv-lock-pre-commit/releases/tag/0.3.0).

> Add an extra hook to check if Pipenv.lock is up-to-date (pipenv verify)

Implemented in version [0.4.0](https://github.com/d-ryzhikov/pipenv-lock-pre-commit/releases/tag/0.4.0).

> Support non-standard Pipenv file locations (via hook argument) for repos with multiple sub-projects

I don't fully understand what you mean by this. Could you please provide an example?


All in all thank you for the suggestions. Feel free to contribute any improvements - I will be glad to review and accept them.

**Comment by AndBondStyle** [comment]
Many thanks for the implemented features, exactly what I wanted!

About "non-standard pipenv file locations". The main use cases I see is monorepos and all-in-one repositories with multiple sub-project or microservices. It means there are multiple `Pipfile` files in multiple sub-directories. You can tell pipenv to use specific pipfile simply by `cd'ing` into directory or by setting [`PIPENV_PIPFILE`](https://pipenv.pypa.io/en/latest/advanced/#pipenv.environments.Setting.PIPENV_PIPFILE). Looking at your code, you can implement it something like that:
```diff
+ from pathlib import Path

def requirements():
    parser = ArgumentParser()
    parser.add_argument(
        "--requirements-file", type=FileType("w"), default="requirements.txt"
    )
    # fetch list of changed files passed by pre-commit
    parser.add_argument("changed_pipfiles", nargs="+")
    args, extra_args = parser.parse_known_args()
+   # get list of pipenv roots (where there is Pipfile OR Pipfile.lock)
+   root_dirs = set(Path(path).absolute().parent for path in args.changed_pipfiles)
+   # run pipenv for each root directory
+   for root_dir in root_dirs:
        subprocess.run(
            ["pipenv", "requirements", *extra_args],
            stdout=args.requirements_file,
+           cwd=root_dir,
        )
```

Quick explanation: pre-commit gives us a list of changed files. If I have 2 subprojects each with their own Pipfile, the hook can receive something like this: [`project_a/Pipfile`, `project_a/Pipfile.lock`, `project_b/Pipfile`]. Generator expression with a set(...) gives us a list of "pipenv roots" - directories containing `Pipfile`. Then, we just call pipenv for them separately, using `cwd` argument in subprocess.run(...).

I'm aware that it's a very niche feature, not sure if someone needs it besides me :) If I'll have some spare time, I can try to implement and test this feature fully and make a PR for you. Implementing it right now is not critical for me at all.

**Comment by d-ryzhykau** [comment]
Implemented monorepo support in [0.5.0](https://github.com/d-ryzhykau/pipenv-lock-pre-commit/releases/tag/0.5.0).

**Comment by AndBondStyle** [comment]
Looks good, thanks! 👍
